sdk/client.go

```diff
@@ -0,0 +1,186 @@
+	"fmt"
+	"io"
+	"mime/multipart"
+	"net/http"
+	"net/url"
+	"sync"
+
+	"github.com/root-gg/plik/server/common"
+)
+
+var (
+	userAgent = "Plik Golang SDK v1.0.0"
+)
+
+// Client is the main plik client
+// It will allow to intercact with plik API to :
+//  - Get/Create/Delete uploads
+//  - Add/Remove files to the upload
+//  - Get plik server configuration
+type Client struct {
+	HttpClient *http.Client
+	BaseURL    *url.URL
+	serverConf *common.Configuration
+	once       *sync.Once
+}
+
+// NewClient instantiate a new Plik client from an URL
+func NewClient(baseURL string) (c *Client, err error) {
+
+	c = new(Client)
+	c.HttpClient = new(http.Client)
+	c.BaseURL, err = url.Parse(baseURL)
+	c.once = new(sync.Once)
+	if err != nil {
+		return
+	}
+
+	return
+}
+
+// NewUpload creates a new empty upload on Plik
+func (c *Client) NewUpload() (upload *Upload) {
+	return c.NewUploadWithOptions(&UploadOptions{})
+}
+
+// NewUploadWithOptions creates a new empty upload on Plik
+// with custom options such as TTL, OneShot, Removable,...
+func (c *Client) NewUploadWithOptions(opts *UploadOptions) (upload *Upload) {
+
+	upload = new(Upload)
+	upload.PlikUpload = common.NewUpload()
+	upload.Files = make(map[string]*File)
+	upload.client = c
+	upload.TTL = opts.TTL
+	upload.Comments = opts.Comments
+	upload.Stream = opts.Stream
+	upload.OneShot = opts.OneShot
+	upload.Removable = opts.Removable
+	upload.ProtectedByPassword = opts.ProtectedByPassword
+	upload.User = opts.User
+	upload.Token = opts.Token
+	upload.Login = opts.Login
+	upload.Password = opts.Password
+	upload.ProtectedByYubikey = opts.ProtectedByYubikey
+	upload.Yubikey = opts.Yubikey
+	upload.IsAdmin = true
+
+	return
+}
+
+// GetUpload will get upload from Plik
+func (c *Client) GetUpload(uploadId string) (upload *common.Upload, err error) {
+
+	path := fmt.Sprintf("/upload/%s", uploadId)
+	req, err := c.newRequest("GET", path, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	_, err = c.do(req, &upload)
+	if err != nil {
+		return nil, err
+	}
+
+	return upload, err
+}
+
+//
+// Private subs
+//
+
+func (c *Client) newRequest(method, path string, body interface{}) (req *http.Request, err error) {
+
+	rel := &url.URL{Path: path}
+	u := c.BaseURL.ResolveReference(rel)
+
+	// Encode body if present
+	var buf io.ReadWriter
+	if body != nil {
+		buf = new(bytes.Buffer)
+
+		err := json.NewEncoder(buf).Encode(body)
+		if err != nil {
+			return nil, err
+		}
+
+	}
+
+	// Make request
+	req, err = http.NewRequest(method, u.String(), buf)
+	if err != nil {
+		return nil, err
+	}
+
+	req.Header.Set("Content-Type", "application/json")
+	req.Header.Set("Accept", "application/json")
+	req.Header.Set("User-Agent", userAgent)
+
+	return
+}
+func (c *Client) newFileRequest(method, path string, fileName string, file io.Reader) (req *http.Request, err error) {
+
+	rel := &url.URL{Path: path}
+	u := c.BaseURL.ResolveReference(rel)
+
+	// Check filename
+	if fileName == "" {
+		fileName = "file"
```

Comment:
Why not use real file name in this case ? Or at least define this default value as a var like user agent ?

Reply:
A default value is a good idea, I'll add it.